server/lib/games/game-loader.js

```diff
@@ -0,0 +1,185 @@
+
+function generateSeed() {
+  // BWChart and some other replay sites/libraries utilize the random seed as the date the game was
+  // played, so we match BW's random seed method (time()) here
+  return (Date.now() / 1000) | 0
+}
+
+function createRoutes(players) {
+  // Generate all the pairings of players to figure out the routes we need
+  const matchGen = []
+  let rest = players
+  while (!rest.isEmpty()) {
+    const first = rest.first()
+    rest = rest.rest()
+    if (!rest.isEmpty()) {
+      matchGen.push([first, rest])
+    }
+  }
+  const needRoutes = matchGen.reduce((result, [p1, players]) => {
+    players.forEach(p2 => result.push([p1, p2]))
+    return result
+  }, [])
+  const pingsByPlayer = new Map(players.map(player => [player, pingRegistry.getPings(player.name)]))
+
+  const routesToCreate = needRoutes.map(([p1, p2]) => ({
+    p1,
+    p2,
+    server: pickServer(pingsByPlayer.get(p1), pingsByPlayer.get(p2)),
+  }))
+
+  return Promise.all(
+    routesToCreate.map(({ p1, p2, server }) =>
+      server === -1
+        ? Promise.reject(new Error('No server match found'))
+        : routeCreator.createRoute(pingRegistry.servers[server]).then(result => ({
+            p1,
+            p2,
+            server: pingRegistry.servers[server],
+            result,
+          })),
+    ),
+  )
+}
+
+const LoadingData = new Record({
+  players: new Set(),
+  finishedPlayers: new Set(),
+  cancelToken: null,
+  deferred: null,
+})
+
+export const LoadingDatas = {
+  isAllFinished(loadingData) {
+    return loadingData.players.every(p => loadingData.finishedPlayers.has(p.id))
+  },
+}
+
+export class GameLoader {
+  constructor() {
+    this.loadingGames = new Map()
+  }
+
+  // A function that starts the process of loading a new game. The first argument is a list of
+  // players, that should be created as a human (or observer) type slots. At least one player should
+  // be present for things to work properly. Besides players, a set of handlers can be sent, which
+  // will be called during various phases of the loading process.
+  //
+  // Returns a promise which will resolve with the list of players if the game successfully loaded,
+  // or rejected if the load failed.
+  loadGame(players, onGameSetup, onRoutesSet) {
+    const cancelToken = new CancelToken()
+    const gameId = cuid()
+    const gameLoad = this._doGameLoad(gameId, cancelToken, players, onGameSetup, onRoutesSet)
+
+    rejectOnTimeout(gameLoad, GAME_LOAD_TIMEOUT).catch(() => {
+      cancelToken.cancel()
+      this.maybeCancelLoading(gameId)
+    })
+
+    return gameLoad
+  }
+
+  // A function which is called to register the (now loaded) game into the system for each player.
+  // Currently, it only cleans up the loading game state, but in future it will be used to accept
+  // the game results and perhaps registering the user to the gameplay activity registry, if not
+  // already registered.
+  registerGame(gameId, playerName) {
```

Comment:
this naming feels a bit odd to me given that the information passed in is something that this service already "knows", it's just registering the fact that the loading completed. Maybe `registerGameAsLoaded` or something along those lines?